src/fparser/two/utils.py

```diff
@@ -219,6 +219,8 @@ def __new__(cls, string, parent_cls=None):
             # restore readers content when no match is found.
             try:
                 result = cls.match(string)
+            except IndexError:
```

Comment:
Could you explain why this is needed. It seems to me that we should be catching any index errors within the classes themselves.